src/core/events/handler.py

```diff
@@ -57,29 +58,66 @@ def subscribe_event(self, event, hook=None, predicate=None):
         if HunterBase in hook.__mro__:
             self.all_hunters[hook] = hook.__doc__
 
-        if hook not in self.hooks[event]:
+        # registering filters
+        if EventFilterBase in hook.__mro__:
+            if hook not in self.filters[event]:
+                self.filters[event].append((hook, predicate))
+                logging.debug('{} filter subscribed to {}'.format(hook, event))
+
+        # registering hunters
+        elif hook not in self.hooks[event]:
             self.hooks[event].append((hook, predicate))
             logging.debug('{} subscribed to {}'.format(hook, event))
 
+    def apply_filter(self, event):
+        # filters a given event, if the event should be removed, returns True, 
+        status = None
+        for hooked_event in self.filters.keys():
+            if hooked_event in event.__class__.__mro__:
+                for filter_hook, predicate in self.filters[hooked_event]:
+                    if predicate and not predicate(event):
+                        continue
+
+                    logging.debug('Event {} got filtered with {}'.format(event.__class__, filter_hook))
+                    status = filter_hook(event).execute()
+        return status
+
     # getting instantiated event object
     def publish_event(self, event, caller=None):
-        logging.debug('Event {} got published with {}'.format(event.__class__, event))
+        remove_event = False
+        apply_filter_original_code = self.apply_filter.__func__.__code__
+
+        # setting event chain
+        if caller:
+            event.previous = caller.event
+            event.hunter = caller.__class__
+
+        # publishing to subscribers
         for hooked_event in self.hooks.keys():
+            if remove_event:
+                break
             if hooked_event in event.__class__.__mro__:
                 for hook, predicate in self.hooks[hooked_event]:
                     if predicate and not predicate(event):
                         continue
 
-                    if caller:
-                        event.previous = caller.event
-                        event.hunter = caller.__class__
-
+                    # will run once for each event
```

Comment:
If you only want to run the filter once per event, couldn't the filtering test be done in the loop outside of this loop? Then I don't think you would have to mess with `self.apply_filter.__func__.__code__`

Reply:
@lizrice You're right, that is a simpler solution. but if we run the filter before the loop, in the worst case, where the event is not published, the loop will execute anyways, this way, we make sure the filtering is done only if the event is published, so this is for optimization.

Either case, I'm working on a better solution. will commit it